rmf_site_editor/src/widgets/console.rs

```diff
@@ -0,0 +1,471 @@
+        let subscriber = subscriber.with(filter_layer);
+        #[cfg(not(target_arch = "wasm32"))]
+        {
+            let fmt_layer = tracing_subscriber::fmt::Layer::default();
+            let subscriber = subscriber.with(fmt_layer);
+            tracing::subscriber::set_global_default(subscriber);
+        }
+        #[cfg(target_arch = "wasm32")]
+        {
+            tracing::subscriber::set_global_default(subscriber);
+        }
+
+        Self {
+            log_history: Vec::new(),
+            current_log: None,
+            category_filter: filter_hashmap,
+            checked_all: true,
+            stored_checked_all: true,
+            display_limit: 100,
+            show_full_history: false,
+            category_count: vec![0; 4], // for Status, Warning, Error, Bevy
+            receiver: rx_2,
+        }
+    }
+}
+
+impl LogHistory {
+    pub fn copy_log_history(&self) -> String {
+        let mut output_string = String::new();
+
+        for log in &self.log_history {
+            if *self.category_filter.get(&log.category).unwrap() {
+                output_string.push_str(&log.category.to_string());
+                output_string.push_str(&log.message);
+                output_string.push_str("\n");
+            }
+        }
+        output_string
+    }
+
+    pub fn current_log(&self) -> &Option<Log> {
+        &self.current_log
+    }
+
+    pub fn log_history(&self) -> &Vec<Log> {
+        &self.log_history
+    }
+
+    pub fn checked_all_mut(&mut self) -> &mut bool {
+        // If "All" checkbox is freshly clicked
+        if self.stored_checked_all != self.checked_all {
+            self.stored_checked_all = self.checked_all;
+            if self.checked_all {
+                for (cat, present) in &mut self.category_filter {
+                    *present = true;
+                }
+            }
+        }
+        &mut self.checked_all
+    }
+
+    pub fn category_present(&self, category: &LogCategory) -> &bool {
+        self.category_filter.get(category).unwrap()
+    }
+
+    pub fn category_present_mut(&mut self, category: LogCategory) -> &mut bool {
+        self.category_filter.get_mut(&category).unwrap()
+    }
+
+    pub fn displayed_category_count(&mut self) -> usize {
+        let mut total_count: usize = 0;
+        for (category, present) in &self.category_filter {
+            if *present {
+                total_count += &self.category_count[*category as usize];
+            }
+        }
+        total_count
+    }
+
+    pub fn display_limit(&self) -> &usize {
+        &self.display_limit
+    }
+
+    pub fn display_limit_mut(&mut self) -> &mut usize {
+        &mut self.display_limit
+    }
+
+    pub fn receive_log(&mut self) {
+        match self.receiver.try_recv() {
+            Ok(msg) => self.append_log(msg),
+            Err(TryRecvError::Disconnected) => println!("Unable to receive log: Disconnected"),
+            Err(TryRecvError::Empty) => (),
+        }
+    }
```

Comment:
It seems the log is updated once per frame with one message, which means that if, let' say, 30 messages were received in one frame and we were running at 30 FPS, the logs would be delayed by one second.

What do you think about, instead of consuming just one entry, we consume all the events that are currently available?
The main drawback would be that if we receive 10 events in one frame only the last one would be readable in the top widget, but on the other hand we wouldn't have these long delays if there happen to be a lot of events (which mostly happen at startups).

Reply:
You're right, that should be the case. I've modified `receive_log()` to receive all available messages in the channel in [5a040a9](https://github.com/open-rmf/rmf_site/pull/121/commits/5a040a9e2ab9bce509df3798a13059d8ca948d6e)